### ISSUE TITLE
The agent should debounce observed unit status changes before reporting them to Fleet server

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent#1946]
As part of the solution in https://github.com/elastic/elastic-agent/issues/1913 the Beats will now restart automatically whenever the output configuration changes. This may cause the Beats to appear unhealthy to the agent for a brief period of time. 

It is possible that this temporary unhealthy state will be reported to Fleet server if it aligns with the time of the next Fleet checkin resulting in an agent that appears unhealthy for up to 5 minutes (the Fleet checkin long poll duration) even though the agent recovered from this state quickly.

We should debounce status changes reported by units to ensure we only report persistent problems to Fleet. Only reporting status changes to Fleet server if they have persisted for ~30 seconds or so seems like a simple starting point.

Originally discussed in https://github.com/elastic/elastic-agent/issues/1913#issuecomment-1346879176

**Comment by blakerouse** [comment]
We also need to update the Fleet gateway code to disconnect and re-connect to the longpoll with the latest status information once it changes.

Elastic Agent has never done any of this before and it was always possible for Elastic Agent to report an unhealthy status for the 5 minute long poll. That is no different than what is in 8.5. For that reason I believe this to be a feature and we should target it to 8.7.

**Comment by cmacknz** [comment]
Agreed, writing this I did notice I was describing a feature. I'll retarget it to 8.7 and we can pull it into 8.6.1 if we have the time and it seems necessary.

**Comment by cmacknz** [comment]
> We also need to update the Fleet gateway code to disconnect and re-connect to the longpoll with the latest status information once it changes.

We discussed this further, and there is a concern that more frequent disconnects and reconnects will have a negative effect in large scale deployments. There is actually an effort underway to increase the long polling timeout significantly to reduce the work fleet-server needs to do when authenticating a new agent connection.

Our conclusion is that we need to address the agent health changing rapidly from healthy to failed to healthy again on an intentional restart in another way. Ideally in a way that does not filter out status changes for unintentional restarts.

One way to get the best of both worlds (infrequent reconnections, immediate status updates to Fleet) would be to switch from long polling to a Websocket. We also may need to revisit how agent health is presented in the UI, as the status changes for individual units do not necessarily need to be reflected in the global agent health we currently present.

**Comment by cmacknz** [comment]
Closing this, the ideal fix for this is to use a streaming protocol to talk to fleet server.

**Comment by joshdover** [comment]
In the meantime, could we debounce, but with a really long debounce window? This would allow us to increase the long polling to 30m, but still go ahead and add re-checkin behavior now to avoid the problem where the agent may appear unhealthy for 30m unnecessarily. If we made the debounce something as long as 5m, we could at least use 30m timeouts most of the time, without making the current staleness any worse than it already is.

**Comment by blakerouse** [comment]
Do we really want to leave something in an Unhealthy state for 5 minutes, or leave something marked Healthy for 5 minutes when it actually has an issue? Don't we want to show this information clearly with in a reasonable amount of time?

I would think a window of a 1 minute would be a better amount of time. In the case that something is down temporary 1 minute seems like enough time for it to come back and we should update the health status in Fleet. If something is wrong because of configuration that is likely that issue is sticky so once reporting failure it doesn't need to report that same failure again for the 30 minute long-poll, until a new policy is sent to the Elastic Agent which will result in a reconnect of the longpoll when that happens.

**Comment by joshdover** [comment]
To be clear, I'm all for making updating the health info in Fleet as fast as we possibly can. I also want to make sure that our scale targets and resource requirements don't regress when we make this change.

I think what we should do is add a setting to horde to be able to emulate this kind of noise in our scale tests. This would allow us to just run a test with some reasonable guesses for a typical and worst case scenario and see how the system holds up.

I wouldn't be comfortable with making the window shorter than 5 minutes until then, since 5 minute checkins is what all of our tests have been based on thus far.

**Comment by cmacknz** [comment]
I wrote up https://github.com/elastic/elastic-agent/issues/2169 for unifying the agent and Horde, I agree this is a good idea and the drift between the two has and probably will continue to cause issues for us since we aren't testing the actual agent code as deployed by real users.

**Comment by joshdover** [comment]
Sounds like we're aligned. Should we update the issue description with this plan outlined to solve #2169 and validate this change in horde first?

**Comment by cmacknz** [comment]
I reworded this issue to be about finding the lower bound for how fast we can check in when the status changes. I don't think we need #2169 for this but it would help.

Do we want a separate issue for specifically implementing and assessing the impact of a 5m minimum bound on how often the agent can check as described in https://github.com/elastic/elastic-agent/issues/1946#issuecomment-1381925175? That might be the first thing to do since we have specific numbers in mind, and we may want to track it separately since I believe it will be a UI regression if we move to 30m check in intervals without it.

**Comment by jlind23** [comment]
@pierrehilbert updating it to a P1 as this is a blocker for our scalability effort. @joshdover and @cmacknz will discuss the path forward here.

**Comment by joshdover** [comment]
@jlind23 as Craig suggested, I've opened a separate issue for this first change: https://github.com/elastic/elastic-agent/issues/2257

I will move this issue out for now.
